### PR TITLE
Fix error handling in psa_driver_wrapper_xxx_hash_get_num_ops

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

Fix broken error handling in psa_driver_wrapper_xxx_hash_get_num_ops

Does this count as an API break?

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not in 2.28
- [x] **tests** covered by existing

### CODE REVIEW

File: scripts/data_files/driver_templates/psa_crypto_driver_wrappers.h.jinja [hunk 1]
```diff
@@ -472,17 +472,19 @@ static inline psa_status_t psa_driver_wrapper_verify_hash(
     }
 }
 
-static inline uint32_t psa_driver_wrapper_sign_hash_get_num_ops(
-    psa_sign_hash_interruptible_operation_t *operation )
+static inline int psa_driver_wrapper_sign_hash_get_num_ops(
```
Comment (paul-elliott-arm) [inline scripts/data_files/driver_templates/psa_crypto_driver_wrappers.h.jinja:475]: I think this should probably be psa_status_t

File: scripts/data_files/driver_templates/psa_crypto_driver_wrappers.h.jinja [hunk 2]
```diff
@@ -495,17 +497,19 @@ static inline uint32_t psa_driver_wrapper_sign_hash_get_num_ops(
     return( PSA_ERROR_INVALID_ARGUMENT );
 }
 
-static inline uint32_t psa_driver_wrapper_verify_hash_get_num_ops(
-    psa_verify_hash_interruptible_operation_t *operation )
+static inline int psa_driver_wrapper_verify_hash_get_num_ops(
```
Comment (paul-elliott-arm) [inline scripts/data_files/driver_templates/psa_crypto_driver_wrappers.h.jinja:500]: likewise psa_status_t

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

File: library/psa_crypto.c [hunk 3]
```diff
@@ -3601,8 +3603,9 @@ psa_status_t psa_verify_hash_complete(
     status = psa_driver_wrapper_verify_hash_complete(operation);
 
     /* Update ops count with work done. */
-    operation->num_ops = psa_driver_wrapper_verify_hash_get_num_ops(
-        operation);
+    if (status == PSA_SUCCESS) {
```
Comment (paul-elliott-arm) [inline library/psa_crypto.c:3606]: status could potentially be PSA_OPERATION_INCOMPLETE here as well as success. We should likely do this call regardless of status at this point, but only overwrite status if psa_driver_wrapper_verify_hash_get_num_ops() fails.

File: library/psa_crypto.c [hunk 4]
```diff
@@ -3470,7 +3470,9 @@ psa_status_t psa_sign_hash_complete(
                                                    signature_length);
 
     /* Update ops count with work done. */
-    operation->num_ops = psa_driver_wrapper_sign_hash_get_num_ops(operation);
+    if (status == PSA_SUCCESS) {
```
Comment (paul-elliott-arm) [inline library/psa_crypto.c:3473]: See below, we should call `psa_driver_wrapper_sign_hash_get_num_ops()` anyway and only overwrite status on failure.

**Review by paul-elliott-arm — DISMISSED** [review]

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I don't understand why we should make this interface change.

`get_num_ops` is a data accessor. We currently have it as a simple function that returns the value and cannot fail. I'm not sure if it's the best design, but it's not obviously broken.

The current code is broken — `psa_driver_wrapper_sign_hash_get_num_ops` (etc.) returns `PSA_ERROR_INVALID_ARGUMENT` in a context that expects an ops count. But that can only happen if there has been data corruption and `operation->id` is not a valid value. Since we don't assert, we can just treat this can't-happen as a can't-happen and continue on (returning 0).

**Comment by daverodgman** [comment]
> But that can only happen if there has been data corruption and `operation->id` is not a valid value. Since we don't assert, we can just treat this can't-happen as a can't-happen and continue

A hardware accelerated implementation might be capable of failing? Obviously if we are confident that failure isn't possible for any driver, then we can remove returning a status from these functions and go back to the old interface.

**Comment by gilles-peskine-arm** [comment]
> A hardware accelerated implementation might be capable of failing?

Sure, but it would fail during an operation function. `num_ops` is data which the driver is supposed to keep updated. If we were ok with using transparent structures (which we avoid in order to maintain an abstraction barrier), we'd pass the driver the address of the operation structure and instruct it to update the `num_ops` field.

**Comment by daverodgman** [comment]
OK, if we're agreed that there is no issue with not being able to return an error code, then let's go for the simple fix.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**